src/reflect/value.go

```diff
@@ -686,23 +686,67 @@ func (v Value) MapIndex(key Value) Value {
 	panic("unimplemented: (reflect.Value).MapIndex()")
 }
 
+//go:linkname hashmapNewIterator runtime.hashmapNewIterator
+func hashmapNewIterator() unsafe.Pointer
+
+//go:linkname hashmapNext runtime.hashmapNextUnsafePointer
+func hashmapNext(m unsafe.Pointer, it unsafe.Pointer, key, value unsafe.Pointer) bool
+
 func (v Value) MapRange() *MapIter {
-	panic("unimplemented: (reflect.Value).MapRange()")
+	if v.Kind() != Map {
+		panic(&ValueError{Method: "MapRange", Kind: v.Kind()})
+	}
+
+	return &MapIter{
+		m:   v,
+		it:  hashmapNewIterator(),
+		key: New(v.typecode.Key()),
+		val: New(v.typecode.Elem()),
+	}
 }
 
 type MapIter struct {
+	m   Value
+	it  unsafe.Pointer
+	key Value
+	val Value
+
+	valid bool
+	done  bool
 }
 
 func (it *MapIter) Key() Value {
-	panic("unimplemented: (*reflect.MapIter).Key()")
+	if !it.valid {
+		panic("MapIter: Key() called before Next()")
+	}
+
+	if it.done {
+		panic("MapIter.Key called on exhausted iterator")
+	}
+
+	return it.key.Elem()
 }
 
 func (it *MapIter) Value() Value {
-	panic("unimplemented: (*reflect.MapIter).Value()")
+	if !it.valid {
+		panic("MapIter: Value() called before Next()")
+	}
+
+	if it.done {
+		panic("MapIter.Value called on exhausted iterator")
+	}
```

Comment:
You could save a bit of const memory (and sacrifice panic message readability) by collapsing these four panic messages into one:

1. Merge `.valid` and `.done` into one flag that simply indicates whether Key()/Value() are the same.
2. Use the same panic message in both cases (the compiler should merge the two strings).

Not sure it's worth it though. Just an idea.

Reply:
I was trying to match the error messages from upstream Go.